### PR TITLE
[MODEL-10313] update task templates

### PR DESCRIPTION
**Pull request by scottp-ml**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary
Custom task interface classes were added to DRUM, but were only used in the huggingface template.  The interface classes reduce the amount of boilderplate needed for saving models, except in special cases.  They also greatly simplify situations where the model needs custom functions or parameters.  Now that can all be serialized in the CustomTask class.  This allowed removal of the templates related to utilizing a custom class as the estimator.  

## Rationale
Utilizing the interface classes simplifies the templates and improves readability.

### CODE REVIEW

**Comment by scottp-ml** [comment]
jarvis test  test_integration_r_environment

**Comment by scottp-ml** [comment]
jarvis test test_integration_general

**Comment by scottp-ml** [comment]
jarvis test test_integration_general

**Comment by scottp-ml** [comment]
jarvis test test_integration_general

**Comment by yakov-g** [comment]
jenkins run test_integration_general

File: task_templates/1_transforms/1_python_missing_values/custom.py [hunk 1]
```diff
@@ -6,64 +6,47 @@
...
-    # and save it into output_dir so that it can be used later to impute on new data
-    output_dir_path = Path(output_dir)
-    if output_dir_path.exists() and output_dir_path.is_dir():
-        with open("{}/artifact.pkl".format(output_dir), "wb") as fp:
-            pickle.dump(median, fp)
+        # compute medians for all numeric features on training data, store them in a dictionary
+        self.median = X.median(axis=0, numeric_only=True, skipna=True).to_dict()
 
+    def transform(self, data: pd.DataFrame) -> pd.DataFrame:
+        """ This hook defines how DataRobot will use the trained object from fit() to transform new data.
```
Comment (akshoop) [inline task_templates/1_transforms/1_python_missing_values/custom.py:36]: Maybe consider keeping the example mention from the docstring of `transformer` old param, i.e.:
```suggestion
        """
        This hook defines how DataRobot will use the trained object from fit() to transform new data.
        In this example, the trained object is a dictionary with medians per column.
        
```

File: task_templates/1_transforms/3_python3_sklearn_transform/custom.py [hunk 2]
```diff
@@ -9,56 +9,40 @@
...
+        """
+        self.transformer = make_pipeline()
+        self.transformer.fit(X, y)
+
+    def transform(self, X: pd.DataFrame) -> pd.DataFrame:
+        """
+        Parameters
+        ----------
+        X: pd.DataFrame - training data to perform transform on
+        Returns
```
Comment (akshoop) [inline task_templates/1_transforms/3_python3_sklearn_transform/custom.py:40]: Docstring formatting nit:
```suggestion
        X: pd.DataFrame - training data to perform transform on

        Returns
```

File: task_templates/2_estimators/1_python_regression/custom.py [hunk 3]
```diff
@@ -6,68 +6,54 @@
...
-        Returns a dataframe with scored data.
-        In case of regression, score() must return a dataframe with a single column with column name "Predictions".
-    """
-
-    return pd.DataFrame(data=model.predict(data), columns=["Predictions"])
+from datarobot_drum.custom_task_interfaces import RegressionEstimatorInterface
+
+
+class CustomTask(RegressionEstimatorInterface):
+    def fit(self, X: pd.DataFrame, y: pd.Series, **kwargs) -> None:
```
Comment (akshoop) [inline task_templates/2_estimators/1_python_regression/custom.py:16]:
```suggestion
    def fit(self, X: pd.DataFrame, y: pd.Series, row_weights: np.ndarray=None, **kwargs) -> None:
```

File: task_templates/2_estimators/9_python_calibrator/custom.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+Copyright 2021 DataRobot, Inc. and its affiliates.
+All rights reserved.
+This is proprietary source code of DataRobot, Inc. and its affiliates.
+Released under the terms of DataRobot Tool and Utility Agreement.
+"""
+# In some cases, avg(prediction) might not match avg(actuals)
+# This task, added as a calibrator in the end of a regression blueprint, can help to fix that
+# During fit(), it computes and stores the calibration coefficient that is equal to avg(actuals) / avg(predicted) on
+# training data
+# During score(), it multiplies incoming data by the calibration coefficient
```
Comment (akshoop) [inline task_templates/2_estimators/9_python_calibrator/custom.py:11]:
```suggestion
# During predict(), it multiplies incoming data by the calibration coefficient
```

File: task_templates/2_estimators/9_python_calibrator/custom.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+import numpy as np
+import pandas as pd
+
+from datarobot_drum.custom_task_interfaces import RegressionEstimatorInterface
+
+
+class CustomTask(RegressionEstimatorInterface):
+    def fit(self, X: pd.DataFrame, y: pd.Series, **kwargs,) -> None:
+
+        self.multiplier = sum(y) / sum(X[X.columns[0]])
```
Comment (akshoop) [inline task_templates/2_estimators/9_python_calibrator/custom.py:22]: nit:
```suggestion
        self.calibration_coefficient = sum(y) / sum(X[X.columns[0]])
```

File: task_templates/2_estimators/9_python_calibrator/custom.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+    def predict(
+        self, data: pd.DataFrame, **kwargs,  # data that needs to be scored
+    ) -> pd.DataFrame:  # returns scored data
+
+        # This hook defines how DR will use the trained object (stored in the variable `model`) to score new data
+
+        # In case of regression, must return a dataframe with a single column with column name "Predictions"
+        return pd.DataFrame(
+            data=np.array(data[data.columns[0]] * self.multiplier).reshape(-1, 1),
```
Comment (akshoop) [inline task_templates/2_estimators/9_python_calibrator/custom.py:32]: If you use my suggestion above:
```suggestion
            data=np.array(data[data.columns[0]] * self. calibration_coefficient).reshape(-1, 1),
```

File: task_templates/2_estimators/9_python_calibrator/custom.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
+"""
+Copyright 2021 DataRobot, Inc. and its affiliates.
+All rights reserved.
+This is proprietary source code of DataRobot, Inc. and its affiliates.
+Released under the terms of DataRobot Tool and Utility Agreement.
+"""
+# In some cases, avg(prediction) might not match avg(actuals)
```
Comment (akshoop) [inline task_templates/2_estimators/9_python_calibrator/custom.py:7]:
```suggestion
# In some cases, avg(predictions) might not match avg(actuals)
```

**Review by akshoop — APPROVED** [review]
LGTM from Core Modeling with comments, nice refactor

**Review by rvorobii — APPROVED** [review]

**Comment by rvorobii** [comment]
also pinging @yakov-g fyi

**Review by yakov-g — APPROVED** [review]

**Pull request merged**